picard/ui/coverartbox.py

```diff
@@ -103,35 +117,49 @@ def set_data(self, data, force=False, pixmap=None):
             return
 
         cover = self.shadow
-        if self.data:
-            if pixmap is None:
+        if not self.data:
+            self.setPixmap(self.shadow)
+            return
+
+        w, h, displacements = (121, 121, 20)
+        if pixmap is None:
+            if len(self.data) == 1:
                 pixmap = QtGui.QPixmap()
-                pixmap.loadFromData(self.data.data)
-            if not pixmap.isNull():
-                offx, offy, w, h = (1, 1, 121, 121)
-                cover = QtGui.QPixmap(self.shadow)
-                pixmap = pixmap.scaled(w, h, QtCore.Qt.KeepAspectRatio, QtCore.Qt.SmoothTransformation)
-                painter = QtGui.QPainter(cover)
-                bgcolor = QtGui.QColor.fromRgb(0, 0, 0, 128)
-                painter.fillRect(QtCore.QRectF(offx, offy, w, h), bgcolor)
-                x = offx + (w - pixmap.width()) / 2
-                y = offy + (h - pixmap.height()) / 2
-                painter.drawPixmap(x, y, pixmap)
+                pixmap.loadFromData(self.data[0].data)
+            else:
+                stack_width, stack_height = (w + displacements * (len(self.data) - 1), h + displacements * (len(self.data) - 1))
+                pixmap = QtGui.QPixmap(stack_width, stack_height)
+                bgcolor = self.palette().color(QtGui.QPalette.Window)
+                painter = QtGui.QPainter(pixmap)
+                painter.fillRect(QtCore.QRectF(0, 0, stack_width, stack_height), bgcolor)
+                x = w / 2
+                y = h / 2
+                for image in self.data:
+                    thumb = QtGui.QPixmap()
+                    thumb.loadFromData(image.data)
+                    thumb = self.decorate_cover(thumb)
+                    painter.drawPixmap(x - thumb.width() / 2, y - thumb.height() / 2, thumb)
+                    x += displacements
+                    y += displacements
                 painter.end()
-        self.setPixmap(cover)
+                pixmap = pixmap.scaled(w, h, QtCore.Qt.KeepAspectRatio, QtCore.Qt.SmoothTransformation)
+                self.setPixmap(pixmap)
+                pixmap = None
+
+        if pixmap and not pixmap.isNull():
+            cover = self.decorate_cover(pixmap)
+            self.setPixmap(cover)
 
     def set_metadata(self, metadata):
         data = None
         self.related_images = []
         if metadata and metadata.images:
             self.related_images = metadata.images
-            for image in metadata.images:
-                if image.is_front_image():
-                    data = image
-                    break
-            else:
+            log.debug("%s using images:" % (self.name), metadata.images)
```

Comment:
On that, maybe we can cache the pixmap. Since it is updated each time the selection is updated, it can lead to too much wasted processing. Especially since a call to self.data[i].data involves I/O.